source/exe/main_common.cc

```diff
@@ -21,80 +22,112 @@
+  ares_library_init(ARES_LIB_INIT_ALL);
+  Event::Libevent::Global::initialize();
+  RELEASE_ASSERT(Envoy::Server::validateProtoDescriptors());
 
+  switch (options_.mode()) {
+  case Server::Mode::Serve: {
 #ifdef ENVOY_HOT_RESTART
-  std::unique_ptr<Server::HotRestartImpl> restarter;
-  try {
-    restarter.reset(new Server::HotRestartImpl(options));
-  } catch (Envoy::EnvoyException& e) {
-    std::cerr << "unable to initialize hot restart: " << e.what() << std::endl;
-    return 1;
+    if (hot_restart) {
+      restarter_.reset(new Server::HotRestartImpl(options_));
+    }
+#endif
+    if (!hot_restart) {
+      restarter_.reset(new Server::HotRestartNopImpl());
+    }
+
+    Stats::RawStatData::configure(options_);
+    tls_.reset(new ThreadLocal::InstanceImpl);
+    Thread::BasicLockable& log_lock = restarter_->logLock();
+    Thread::BasicLockable& access_log_lock = restarter_->accessLogLock();
+    auto local_address = Network::Utility::getLocalAddress(options_.localAddressIpVersion());
+    Logger::Registry::initialize(options_.logLevel(), log_lock);
+
+    stats_store_.reset(new Stats::ThreadLocalStoreImpl(restarter_->stats_allocator()));
+    server_.reset(new Server::InstanceImpl(options_, local_address, default_test_hooks_,
+                                           *restarter_, *stats_store_, access_log_lock,
+                                           component_factory_, *tls_));
+    break;
   }
+  case Server::Mode::Validate:
+    break;
+  }
+}
 
-  Thread::BasicLockable& log_lock = restarter->logLock();
-  Thread::BasicLockable& access_log_lock = restarter->accessLogLock();
-  Stats::RawStatDataAllocator& stats_allocator = *restarter;
-#else
-  std::unique_ptr<Server::HotRestartNopImpl> restarter;
-  restarter.reset(new Server::HotRestartNopImpl());
+MainCommonBase::~MainCommonBase() { ares_library_cleanup(); }
 
-  Thread::MutexBasicLockable log_lock, access_log_lock;
-  Stats::HeapRawStatDataAllocator stats_allocator;
-#endif
-
-  RELEASE_ASSERT(Envoy::Server::validateProtoDescriptors());
-  Event::Libevent::Global::initialize();
-  Server::ProdComponentFactory component_factory;
-  auto local_address = Network::Utility::getLocalAddress(options.localAddressIpVersion());
-  switch (options.mode()) {
+bool MainCommonBase::run() {
+  switch (options_.mode()) {
   case Server::Mode::Serve:
-    break;
-  case Server::Mode::Validate:
-    Thread::MutexBasicLockable log_lock;
-    Logger::Registry::initialize(options.logLevel(), log_lock);
-    return Server::validateConfig(options, local_address, component_factory) ? 0 : 1;
+    server_->run();
+    return true;
+  case Server::Mode::Validate: {
+    auto local_address = Network::Utility::getLocalAddress(options_.localAddressIpVersion());
+    return Server::validateConfig(options_, local_address, component_factory_);
   }
+  }
+  throw EnvoyException(fmt::format("Invalid server mode: {}", int(options_.mode())));
+  return false;
```

Comment:
I don't think this can actually happen? I would just replace both lines with `NOT_REACHED`

Reply:
correct; I didn't expect it to happen.  that's just a practice I evolved for dealing with untrusted enums coming into switch statements.  NOT_REACHED sgtm.